File: app/src/main/java/net/keysupport/cardread/network/VssService.kt
package net.keysupport.cardread.network

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.kotlinx.serialization.asConverterFactory
import retrofit2.http.Body
import retrofit2.http.POST
import java.util.concurrent.TimeUnit

@Serializable
data class VssRequest(
    val validationPolicyId: String,
    val x509Certificate: String
)

@Serializable
data class ValidationResult(
    val result: String? = null,
    val isAffirmativelyInvalid: Boolean? = null,
    val invalidityReasonText: String? = null
)

@Serializable
data class VssResponse(
    val validationPolicyId: String? = null,
    val x509SubjectName: String? = null,
    val x509CertificatePath: List<String>? = null,
    val error: String? = null,
    val status: String? = null,
    val message: String? = null,
    val validationResult: ValidationResult? = null
)

interface VssApi {
    @POST("vss/v2/validate")
    suspend fun validateCertificate(@Body request: VssRequest): VssResponse
}

object VssClient {
    private val logging = HttpLoggingInterceptor().apply {
        level = HttpLoggingInterceptor.Level.BODY
    }
    
    private val client = OkHttpClient.Builder()
        .addInterceptor(logging)
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .build()
        
    private val json = Json { ignoreUnknownKeys = true }
    
    val api: VssApi by lazy {
        Retrofit.Builder()
            .baseUrl("https://home.keysupport.net/")
            .client(client)
            .addConverterFactory(json.asConverterFactory("application/json".toMediaType()))
            .build()
            .create(VssApi::class.java)
    }
}

File: ios/KeySupportValidator/Network/VSSService.swift
import Foundation

struct VSSRequest: Encodable {
    let validationPolicyId: String
    let x509Certificate: String
}

struct ValidationResult: Decodable {
    let result: String?
    let isAffirmativelyInvalid: Bool?
    let invalidityReasonText: String?
}

struct VSSResponse: Decodable {
    let validationPolicyId: String?
    let x509SubjectName: String?
    let x509CertificatePath: [String]?
    let error: String?
    let status: String?
    let message: String?
    let validationResult: ValidationResult?
}

enum VSSError: Error, LocalizedError {
    case offline
    case cannotConnect
    case timedOut
    case badStatus(Int)
    case malformedResponse

    var errorDescription: String? {
        switch self {
        case .offline:            return "No Internet Connection"
        case .cannotConnect:      return "Could not connect to validation server"
        case .timedOut:           return "Connection to server timed out"
        case .badStatus(let code): return "Validation server returned HTTP \(code)"
        case .malformedResponse:  return "Validation server returned an unreadable response"
        }
    }
}

/// Client for the KeySupport Validation Service.
///
/// Call this only *after* the NFC session has been invalidated. iOS gives a
/// connected tag ~20 seconds, and these timeouts alone could consume most of it.
struct VSSClient {

    /// PIV Card Authentication policy OID.
    static let cardAuthenticationPolicyOID = "2.16.840.1.101.10.2.18.2.2.1"

    private let endpoint = URL(string: "https://home.keysupport.net/vss/v2/validate")!
    private let session: URLSession

    init() {
        let configuration = URLSessionConfiguration.default
        configuration.timeoutIntervalForRequest = 15
        configuration.timeoutIntervalForResource = 30
        configuration.waitsForConnectivity = false
        session = URLSession(configuration: configuration)
    }

    func validate(certificateDER: Data) async throws -> VSSResponse {
        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.httpBody = try JSONEncoder().encode(
            VSSRequest(
                validationPolicyId: Self.cardAuthenticationPolicyOID,
                x509Certificate: certificateDER.base64EncodedString()
            )
        )

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch let urlError as URLError {
            // Map to the same user-facing categories the Android app distinguishes.
            switch urlError.code {
            case .notConnectedToInternet, .dataNotAllowed:
                throw VSSError.offline
            case .cannotConnectToHost, .cannotFindHost, .networkConnectionLost, .dnsLookupFailed:
                throw VSSError.cannotConnect
            case .timedOut:
                throw VSSError.timedOut
            default:
                throw urlError
            }
        }

        if let http = response as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
            throw VSSError.badStatus(http.statusCode)
        }

        do {
            return try JSONDecoder().decode(VSSResponse.self, from: data)
        } catch {
            throw VSSError.malformedResponse
        }
    }
}
